### ISSUE TITLE
Centralized storage of all paths, probably paths.yml

### ISSUE THREAD

**Issue by Boorj** [bolt/bolt#3594]
Make commonly used paths relative, and gather all paths to one file.

All those lines like
```php
require_once $this['resources']->getPath('root/vendor/symfony/locale/Symfony/Component/Locale/Resources/stubs/functions.php');
```
are quite frustrating because of disability to reorganize some folders.

Also weird solution
```php
'monolog.logfile' => $this['resources']->getPath('cache') . '/' . $this['config']->get('general/debuglog/filename')
```
Also translations algorythm has weird directory dependencies
```php
$path = '/resources/translations/' . $locale . '/' . $domain . '.' . $locale . '.yml';
//...
if (strlen($locale) == 5 && !is_dir($this->app['paths']['apppath'] . $path)) { //...
// and if i want to keep my contenttype translations with my project files?
```
![024027](https://cloud.githubusercontent.com/assets/2145313/7867296/fdcc2be6-0596-11e5-8acb-7ad85ef8ff10.jpg)
And all those absolute paths that make loose coupling impossible could be avoided.
![board](https://cloud.githubusercontent.com/assets/2145313/7867330/361ca2aa-0597-11e5-9545-39fdc36297f0.jpg)

**Comment by Boorj** [comment]
Just for example
```
// Application.php
$app['resources']->getPath('root') . '/vendor/symfony/web-profiler-bundle/Symfony/Bundle/WebProfilerBundle/Resources/views'
//to
$app['resources']->getPath('vendor') . '/symfony/web-profiler-bundle/Symfony/Bundle/WebProfilerBundle/Resources/views'
```
or 
```
$path = '/resources/translations/' . $locale . '/' . $domain . '.' . $locale . '.yml'; //...
!is_dir($this->app['paths']['apppath'] . $path))
//to
$path = '/' . $locale . '/' . $domain . '.' . $locale . '.yml'; //...
!is_dir($this->app['paths']['translations'] . $path))
```

**Comment by GawainLynch** [comment]
Yeah, we're still in the process of updating that stuff… but they're being picked up with the refactors

**Comment by Boorj** [comment]
Also it's easiear to get extensions updated. No hard dependencies.

**Comment by GawainLynch** [comment]
Just to cover this off. 

#### Translations Paths
That part will happen, that is down the queue a little bit as we have some bigger issues to address first

#### Vendor Path
Quite simply it is chicken and egg. The vendor path is referred to in `app/bootstrap.php` in order to include the autoloader. Without the autoloader we have no way of loading and processing the YAML files. Others are (obliviously) most welcome to offer better suggestions for a *solution*.

For me, this is a `WONTFIX`.

**Comment by rossriley** [comment]
I'm also against any official way to move the vendor path.

In terms of configuration It's also much nicer to use the configuration object rather than having a separate config file. We designed the config object to allow more runtime flexibility, and of course it makes it much easier for testing where we don't want to have to worry about lots of non-php config to get us setup.

In terms of flexibility though, we can carry on adding more path options to the configuration. 

It's also worth noting that having the configuration as an object means that you can create your own config option, that extends either `Bolt\Configuration\Composer` or `Bolt\Configuration\Standard` and include that as a dependency of your projects.

**Comment by Boorj** [comment]
*2 rossiley*
Again: the symfony guys are ok with vendor dir moving:
http://symfony.com/doc/current/cookbook/configuration/override_dir_structure.html#override-the-vendor-directory

*2 GawainLynch*
yes i know that it's a weird question.
i use this construction
```php
// index.php
$app_dir    = 'app';
$vendor_dir = "{$app_dir}/vendor";
require_once "../{$vendor_dir}/autoload.php";
$configuration = new Bolt\Configuration\Composer(dirname(__DIR__));
$configuration->setPath('vendor_dir',             $vendor_dir);
```

But anyway
1. this issue was **not** about vendor dir movement. It was about one config file for all directories, including thumbnails etc. If it doesn't work for vendor dir - ok. But the rest part? why is it "wontfix"? 
2. i'm event not against keeping vendor dir in %root%/vendor  . 
symfony guys also use this hardcoded line:
https://github.com/symfony/symfony-standard/blob/2.8/app/autoload.php#L9

But why can't it be used here 
```php
// /bolt/.../Application.php
$app['resources']->getPath('root') . '/vendor/symfony/web-profiler-bundle/Symfony/Bundle/WebProfilerBundle/Resources/views'
```
where instead you can use  `getPath('root') . '/vendor'`  instead of  `getPath('vendor_dir')` so it stands relative to one folder that is set up in global config file OR it's being got from autoload.php file automatically like in composer's generated autoloader_namespaces.php : 
```php
$vendorDir = dirname(dirname(__FILE__));
$baseDir = dirname($vendorDir);
return array(
    'Twig_Extensions_'                     => array($vendorDir . '/twig/extensions/lib'),
// ...
```

**Comment by GawainLynch** [comment]
> But the rest part? why is it "wontfix"? 

The "wontfix" was directly aimed at the vendor move.

I think your solution probably lies in a slightly different approach, namely the very outstanding #3005 which (I think) just needs documentation written to be accepted for merge.

**Comment by rossriley** [comment]
Ok, if we add vendor (and any other missing paths ) to the configuration object then does this solve what you need.

We obviously need to do a pass through the code of places where paths are generated, I know there's a few that haven't been converted to the new style yet.

For reference this one

```
// /bolt/.../Application.php
$app['resources']->getPath('root') . '/vendor/symfony/web-profiler-bundle/Symfony/Bundle/WebProfilerBundle/Resources/views'
```
Would become...

```
// /bolt/.../Application.php
$app['resources']->getPath('vendorsymfony/web-profiler-bundle/Symfony/BundleWebProfilerBundle/Resources/views');
```

**Comment by CarsonF** [comment]
^ that instance doesn't even exist anymore. If it does, it should be removed.

**Comment by GawainLynch** [comment]
Those paths are definitely gone… the :koala: ate them :wink:

**Comment by Boorj** [comment]
Again, you're talking about several instances and particular  solutions
my index.php is

```php
// index.php
<?php
$vendor_dir = "app/vendor";
require_once "../{$vendor_dir}/autoload.php";

$app_dir    = 'app';
$src_dir    = 'src';
$web_dir    = 'web';
$upload_dir = "up";

$configuration = new Bolt\Configuration\Composer(dirname(__DIR__));
$configuration->setPath('vendor_dir',             $vendor_dir);
$configuration->setPath('cache',                  "{$app_dir}/vars/cache");
$configuration->setPath('logs',                   "{$app_dir}/vars/logs");
$configuration->setPath('database',               "{$app_dir}/database");
$configuration->setPath('extensionspath',         "{$app_dir}/extensions");
$configuration->setPath("themebase",              "{$src_dir}");
$configuration->setPath('config',                 "{$src_dir}/config");
$configuration->setPath('extensionsconfig',       "{$src_dir}/config/extensions");
$configuration->setPath("web",                    "{$web_dir}");
$configuration->setPath('files',                  "{$web_dir}/{$upload_dir}");


$configuration->getVerifier()->disableApacheChecks();
$configuration->verify();
$app = new Bolt\Application(array('resources'=>$configuration));
$app->initialize();

// thumbs dir is inside upload dir to gitignore just one dir.
$app->mount("/{$upload_dir}/thumbs", new \Bolt\Thumbs\ThumbnailProvider());

$app->register(new SilexMtHaml\MtHamlServiceProvider());
require_once "../src/controllers/Front.php";

$app->run();
```
and it seems like it is possible to do via some `paths.yml` file (probably except of vendor dir).

**Comment by Boorj** [comment]
and also it doesn't ignore the need in global variables used in configuration yamls, suggested in #3005 .

**Comment by CarsonF** [comment]
So you want to define paths in a config file and one of those paths is where the config file is? :chicken: :egg:

**Comment by Boorj** [comment]
Mentioned above: exceptions are accepted. You can call autoload.php directly in index.php and set the vendor directory variable inside autoload.php. And for config.dir variable you can made an exception and define that location directly in index.php.
Actually, good way is to call different front controllers for different environment, so each of them has it's own set of variables. In my case i send request first to app.php/app_dev.php that includes index.php. So all dirs could be set directly there (i know that yaml loader is not loaded at that point, but it could be also solved)

**Comment by rossriley** [comment]
Ok, maybe I'm missing something that's not covered here but just to demonstrate I've put together an example of what I'd do in this situation.

https://gist.github.com/rossriley/a1647629c70a0336df4f

We designed the configuration object so that we could do exactly this, and if this config setup is something you use a lot, you could even package it up separately as its own composer package and require it in all your Bolt projects.

**Comment by CarsonF** [comment]
Does `vendor` need to be defined in the paths? I don't think it does.

**Comment by Boorj** [comment]
for vendor - ok. Actually it's just my own habits not to use absolute paths or do it rarely.

**Comment by Boorj** [comment]
So now my config looks like this:
https://gist.github.com/Boorj/151597e0f071d431a32f

Thanks guys, now it's a nice boilerplate for me.

But anyway i consider one storage for paths - a good practice
```php
    // frustrating...
        $configuration->setPath("cache",                  "{$app_dir}/vars/cache");
        $configuration->setPath("logs",                   "{$app_dir}/vars/logs");
        $configuration->setPath("database",               "{$app_dir}/database");
        $configuration->setPath("extensionspath",         "{$app_dir}/extensions");
        $configuration->setPath("themebase",              "{$res_dir}");
        $configuration->setPath("config",                 "{$res_dir}/config");
        $configuration->setPath("extensionsconfig",       "{$res_dir}/config/extensions");
        $configuration->setPath("translations",           "{$res_dir}/translations");
```

```yml
# ( ͡° ͜ʖ ͡°)  hell yeah))
        cache             :  "%app_dir%/vars/cache"
        logs              :  "%app_dir%/vars/logs"
        database          :  "%app_dir%/database"
        extensionspath    :  "%app_dir%/extensions"
        themebase         :  "%res_dir%"
        config            :  "%res_dir%/config"
        extensionsconfig  :  "%res_dir%/config/extensions"
        translations      :  "%res_dir%/translations"
        web               :  "%web_dir%"
        files             :  "%web_dir%/%upload_dir%"
        thumbs            :  "/%upload_dir%/thumbs"
```

Yes, i know that from this point i can make it by my own, but i thought it would be lovely addition to the Bolt core.

**Comment by GawainLynch** [comment]
Fixed in #5069
